File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        int N = board.length;

        for (int move : moves) {
            for (int i = 0; i < N; i++) {
                if (board[i][move - 1] != 0) {
                    int target = board[i][move - 1];
                    if (check(target, stack)) {
                        answer++;
                    }
                    board[i][move - 1] = 0;
                    break;
                }
            }
        }

        return answer * 2;
    }

    public static boolean check(int target, ArrayDeque<Integer> stack) {
        if (stack.isEmpty()) {
            stack.addLast(target);
            return false;
        }

        if (stack.peekLast() == target) {
            stack.pollLast();
            return true;
        }

        stack.addLast(target);
        return false;
    }
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
def solution(board, moves):
    answer = 0
    b = []
    # moves 만큼 반복하기
    for move in moves:
        # 0행, move 열부터 
        c = move - 1
        for r in range(len(board)):
            if board[r][c] != 0:
                print(board[r][c])
                if not b:
                    b.append(board[r][c])
                else:
                    if b[-1] == board[r][c]:
                        b.pop()
                        answer += 2
                    else:
                        b.append(board[r][c])
                board[r][c] = 0
                break
    return answer